Vecmem main

keyradical opened the pull request:
The code now compiles but the ActsUnitTestSeedfinderSycl gives segmentation fault when running.

krasznaa committed:
Taught Acts how to make use of VecMem in its SYCL plugin.

Using the exact same formalism that we use in traccc as well.

Note that the code is not actually making use of the VecMem types
just yet.

Commit:
marked unit tests with ActsUnitTestSeedfinderSycl

Commit:
included vecmem shared memory resource pointer in main

Commit:
Created vecmem memory resource interface and added it to the constructor of Seedfinder

Commit:
Modified data types in the constructor to vecmem types

krasznaa commented on Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/CreateSeedsForGroupSycl.hpp:
```diff
@@ -36,8 +39,8 @@ void createSeedsForGroupSycl(
-    const std::vector<detail::DeviceSpacePoint>& topSPs,
+    vecmem::vector<detail::DeviceSpacePoint>& bottomSPs,
+    vecmem::vector<detail::DeviceSpacePoint>& middleSPs,
+    vecmem::vector<detail::DeviceSpacePoint>& topSPs,
```
You should've kept the `const` declaration of these vectors. You only use them as inputs in the function, you don't need them to be non-const.

krasznaa commented on Tests/UnitTests/Plugins/Sycl/Seeding/SeedfinderSyclTest.cpp:
```diff
@@ -122,6 +126,7 @@ auto setupSpacePointGridConfig(
 
 auto main(int argc, char** argv) -> int {
   auto start_prep = std::chrono::system_clock::now();
+  vecmem::sycl::shared_memory_resource* resource = nullptr;
```
:stuck_out_tongue: I was expecting something of a bit more elaborate issue. :smile:

It was correct to receive the `vecmem::memory_resource` pointer in `Acts::Sycl::Seedfinder` as you did. That part you did as I imagined.

But then, you need to give a valid pointer to the seedfinder. :wink: The most easily by creating your `vecmem::sycl::shared_memory_resource` on the stack. Like:

```c++
vecmem::sycl::shared_memory_resource resource;
```

This will create an actual object in memory, not just a 64-bit empty `nullptr`. Then, and I guess this is what confused you, you can give this object's pointer to the seedfinder constructor like:

```c++
  Acts::Sycl::Seedfinder<SpacePoint> syclSeedfinder(
      config, deviceAtlasCuts,
      &resource,
      Acts::Sycl::QueueWrapper(
      ...
```

I.e. you can use `&` to give a pointer to this real object to the seedfinder.

Give that a try. :wink:

keyradical replied:
Ooh, obviously, that makes more sense.

krasznaa reviewed (commented):
Once the code runs, we can talk about the next complications in synchronizing how `cl::sycl::queue` is used by the seedfinder code. Since currently the algorithmic code and vecmem use it in different ways, not necessarily making use of the same device in all situations.

Commit:
Modified function inputs to const vecmem types

Commit:
Changed the shared memory resource from nullptr to actual vecmem object

Commit:
created new constructor with queue pointer

Commit:
created queue and linked memory resource to it

Commit:
modified Duplet Search so that it uses VecMem vectors

Commit:
Linked memory resource to CreateSeedsForGroupSycl

Commit:
Created VecMem jagged vectors linked to the memory resource.
Then, created jagged vector views and passed them into Duplet Search.

Commit:
Created jagged device vector from the vector view and pushed back the compatible indices.

Commit:
Got rid of the Atomic accessor type.

Commit:
Fixed the proper initialization of jagged vectors and deleted some unused, previous code.

Commit:
Made use of the VecMem types.

Commit:
Tried changing the code's workflow to be compatible with VecMem.

Commit:
Modified the code up to the end of Linear Transform.
Fixed the logic of handling jagged vectors when creating
the flattend vector of all compatible space points.
This was earlier done for 'easier indexing' inside Linear Transform.
It might not be needed and jagged vectors could be passed directly to LT.

The code is not yet possible to compile but so far all the errors come from later, not yet touched code of Triplet Search.

Commit:
Code modified up to the end to make use of Vecmem vectors but many errors generated.

Commit:
VecMem vectors implemented into the code up to and involving Linear Transform.
- Linear Transform's results are also stored in VecMem vectors
- The code compiles but the Seedfinder Sycl test gives 0 agreement.

Commit:
Revert "Code modified up to the end to make use of Vecmem vectors but many errors generated."

This reverts commit 316f20cb5a6a3fb7f515222819224366850d522b.

Commit:
Merge branch 'old-state' into Vecmem-main

Commit:
Code compiles - VecMem vectors included up to the end of triplet search but only partially in TS.

Commit:
Linked vecmem::sycl and disabled --coverage compiler flag

Commit:
Created jagged vecmem vector buffers and passed them to Duplet Search. it still doesn't execute properly, no duplets found.

Commit:
Made use of the vecmem buffers.
The code compiles and the test executes but the efficiency of
the SYCL plugin is low (slow) and the agreement of found seeds is 0. So not the best.

Commit:
Further improvements in the triplet search.
No luck with improving the efficiency or correctness of the code.

Commit:
further changes
